torch/onnx/_internal/fx/_pass.py

```diff
@@ -161,8 +168,28 @@ def __init__(
             diagnostic_context: The diagnostic context for recording diagnostics.
             module: The module to be transformed.
         """
-        self.module = module
         self.diagnostic_context = diagnostic_context
+        self.module = module
+        self.fake_mode = self._detect_fake_mode()
+
+    def _detect_fake_mode(self) -> Optional[fake_tensor.FakeTensorMode]:
+        """Detect fake mode from the graph.
+
+        Scan through all nodes in graph and their meta['val'] to detect fake mode.
+        """
+        fake_tensors = [node.meta.get("val") for node in self.module.graph.nodes]
+        return torch._dynamo.utils.detect_fake_mode(fake_tensors)
+
+    def _maybe_fakefy_args(
+        self, fake_mode: Optional[fake_tensor.FakeTensorMode], *args: Any
+    ) -> Tuple[Any, ...]:
+        if fake_mode is None:
+            return args
+        # NB: This should hit the cache if tensors were fakefied before.
+        # E.g., when the fx graph is produced by Dynamo.
+        return tuple(
+            fake_mode.from_tensor(t) if isinstance(t, torch.Tensor) else t for t in args
+        )
```

Comment:
Maybe move this to a `fake_tensor.py` or `fake_util.py` file? This is a domain specific util not really related to passes only. It is hard to reuse a generic util that is hidden into an unrelated file

This is used by `_pass.py` and maybe (see comment below) should be used by `type_promotion.py`; it might be used at another components when dynamic shape is fully supported

```suggestion
    def _detect_fake_mode(fx_module) -> Optional[fake_tensor.FakeTensorMode]:
        """Detect fake mode from the graph.

        Scan through all nodes in graph and their meta['val'] to detect fake mode.
        """
        fake_tensors = [node.meta.get("val") for node in self.module.graph.nodes]
        return torch._dynamo.utils.detect_fake_mode(fake_tensors)

    def _maybe_fakefy_args(
        fake_mode: Optional[fake_tensor.FakeTensorMode], *args: Any
    ) -> Tuple[Any, ...]:
        if fake_mode is None:
            return args
        # NB: This should hit the cache if tensors were fakefied before.
        # E.g., when the fx graph is produced by Dynamo.
        return tuple(
            fake_mode.from_tensor(t) if isinstance(t, torch.Tensor) else t for t in args
        )
```

Reply:
Similar to reply at https://github.com/pytorch/pytorch/pull/105764#discussion_r1271612064, but I will do this in a follow up.